client/app/pages/queries/view.js

```diff
@@ -157,7 +157,7 @@ function QueryViewCtrl(
   $scope.canEdit = currentUser.canEdit($scope.query) || $scope.query.can_edit;
   $scope.canViewSource = currentUser.hasPermission('view_source');
 
-  $scope.canExecuteQuery = () => currentUser.hasPermission('execute_query') && !$scope.dataSource.view_only;
+  $scope.canExecuteQuery = () => $scope.query.is_safe || (currentUser.hasPermission('execute_query') && !$scope.dataSource.view_only);
```

Comment:
What if the user changes the query and it's no longer safe?

Reply:
Then he will [get a 403 back](https://github.com/getredash/redash/pull/3383/files#diff-f734d32676ec526e17e5ae92d41b5787R195) from the server. I'll say this is an edge case, which while not handled in the most elegant way, is a good enough solution.

We might want to consider a better error message there, though.

Reply:
I think we should prevent editing query source when the user doesn't have Full Access permission to the data source. We might actually be doing this already.

Then this becomes a non issue. We can open a separate issue for this.